README
                                                           -*- Text -*-
This is the README file for Emacs Muse.

The terms "Emacs Muse", "Emacs-Muse", and "Muse" are used
interchangeably throughout this document to refer to this software.

Synopsis
========

Emacs Muse is an authoring and publishing environment for Emacs.  It
simplifies the process of writings documents and publishing them to
various output formats.

Emacs Muse consists of two main parts: an enhanced text-mode for
authoring documents and navigating within Muse projects, and a set of
publishing styles for generating different kinds of output.

Directory contents
==================

The directory structure is as follows.

etc :: Miscellaneous files like the emacs-wiki migration guide, the
  list of ideas for the future, and a RelaxNG schema for Muse XML.

contrib :: Files that are not a part of Muse proper, but are useful
  enough to be bundled with Muse.

examples :: Example configuration files for use with Muse, a
  publishable quickstart guide, and miscellaneous helper stuff.

experimental :: Source code that is not yet considered stable.

lisp :: Muse source code.

scripts :: Scripts that are used when compiling Muse and publishing
  the QuickStart document; they might prove useful if you want to
  automate the building of Muse documents.

texi :: The manual (muse.texi) and any supplemental sections.

Getting started
===============

Prerequisites
-------------

You need either Emacs (21.1 or greater), XEmacs (21.4 or greater), or
SXEmacs (warning: Muse has not been tested with SXEmacs).  The current
maintainer recommends using Emacs.

Compilation
-----------

This is an optional step, since Emacs Lisp source code does not
necessarily have to be byte-compiled.  It may yield a slight speed
increase, however.

A working copy of Emacs or XEmacs is needed in order to compile the
Emacs Muse.  By default, the program that is installed with the name
`emacs' will be used.

If you want to use the `xemacs' binary to perform the compilation, you
must copy `Makefile.defs.default' to `Makefile.defs' in the top-level
directory, and then edit `Makefile.defs' as follows.  You can put
either a full path to an Emacs or XEmacs binary or just the command
name, as long as it is in the PATH.  Depending on your setup, changes
to the PREFIX, ELISPDIR, and/or INFODIR variables may also need to be
made.

EMACS    = xemacs
SITEFLAG = -no-site-file
# Edit the section as necessary
install_info = install-info --section "XEmacs 21.4" $(1).info \
        $(INFODIR)/dir || :

Running `make' in the top-level directory should compile the Muse
source files in the `lisp' directory, and generate an autoloads
file in `lisp/muse-autoloads.el'.

Installation
------------

Muse may be installed into your file hierarchy by doing the following.

Copy `Makefile.defs.default' to `Makefile.defs' in the top-level
directory, if you haven't done so already.  Then edit the
`Makefile.defs' file so that ELISPDIR points to where you want the
source and compiled Muse files to be installed and INFODIR indicates
where to put the Muse manual.  As mentioned earlier, you will want to
edit EMACS and SITEFLAG as shown in the Compilation section if you are
using XEmacs.

If you are installing Muse on a Debian or Ubuntu system, you might
want to change the value of INSTALLINFO as specified in
`Makefile.defs'.

If you wish to install Muse to different locations than the defaults
specify, edit `Makefile.defs' accordingly.

Run `make' as a normal user, if you haven't done so already.

Run `make install' as the root user if you have chosen installation
locations that require this.

Insinuation
-----------

Two things need to happen in order for Muse to be usable with your
version of Emacs or XEmacs.

 1. The location of the Muse directory needs to be added to the load
    path so that your variant of Emacs can find it.

 2. You need to load whichever Muse files you wish to make use of.

A quick example that accomplishes both of these follows.

;; Add this to your .emacs or .xemacs/init.el file.
(setq load-path (add-to-list 'load-path "/path/to/muse"))
(require 'muse-mode)
(require 'muse-publish)
(require 'muse-html)  ;; and so on

Documentation
-------------

The Muse manual may be generated by running `make doc'.

It will produce two files: an Info file called `muse.info' and an HTML
document called `muse.html'.  Both of these will be created in the
`texi' directory.

This manual is also available online in several forms.

 - PDF: http://mwolson.org/static/doc/muse.pdf
 - HTML (single file): http://mwolson.org/static/doc/muse.html
 - HTML (multiple files): http://mwolson.org/static/doc/muse/

QuickStart
----------

Type "make examples" to generate a quickstart guide for Muse.  An HTML
document called QuickStart.html, an Info document called
QuickStart.info, and a PDF file called QuickStart.pdf will be created
in the `examples' directory.

If you do not have a working LaTeX installation with the proper fonts,
the PDF file will not be successfully generated.

To see the document that is interpreted in order to generate these
files, take a look at `examples/QuickStart.muse'.

To view the generated Info document from within Emacs, try the
following.

C-u M-x info RET /path/to/muse/examples/QuickStart.info RET

Further Documentation
=====================

Please consult http://www.emacswiki.org/cgi-bin/wiki/EmacsMuse to find
more information on bugs, unimplemented features, and user-contributed
tweaks.

The current maintainer's webpage for Muse is at
http://mwolson.org/projects/EmacsMuse.html.

If you wish to participate in the development of Muse, or track the
latest development changes, please read the `Development' section of
the Muse manual (in the `Obtaining Muse' chapter) for instructions.

Mailing Lists
=============

Mailing lists for Muse exist.

muse-el-announce ::
  Low-traffic list for Muse-related announcements.

  You can join this mailing list (muse-el-announce@gna.org)
  using the subscription form at
  http://mail.gna.org/listinfo/muse-el-announce/.

  This mailing list is also available via Gmane (http://gmane.org/).
  The group is called
  gmane.emacs.muse.announce.

muse-el-discuss ::
  Discussion, bugfixes, suggestions, tips, and the like for Muse.
  This mailing list also includes the content of muse-el-announce.

  You can join this mailing list (muse-el-discuss@gna.org)
  using the subscription form at
  http://mail.gna.org/listinfo/muse-el-discuss/.

  This mailing list is also available via Gmane with the identifier
  gmane.emacs.muse.general.

muse-el-logs ::
  Log messages for commits made to Muse.

  You can join this mailing list (muse-el-logs@gna.org) using
  the subscription form at
  http://mail.gna.org/listinfo/muse-el-logs/.

  This mailing list is also available via Gmane with the identifier
  gmane.emacs.muse.scm.

muse-el-commits ::
  Generated bug reports for Emacs Muse.  If you use our bug-tracker at
  https://gna.org/bugs/?group=muse-el, the bug reports will be sent to
  this list automatically.

  You can join this mailing list (muse-el-commits@gna.org) using
  the subscription form at
  http://mail.gna.org/listinfo/muse-el-commits/.

  This mailing list is also available via Gmane with the identifier
  gmane.emacs.muse.cvs.

License
=======

Emacs Muse is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 3, or (at your option) any
later version.

Emacs Muse is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with Emacs Muse; see the file COPYING.  If not, write to the
Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301, USA.

The Muse manual is available under the terms of the GNU Free
Documentation License, Version 1.2 or any later version published by
the Free Software Foundation; with no Invariant Sections, and with the
Front-Cover texts and Back-Cover Texts as specified in the manual.  A
copy of the license is included in the section entitled "GNU Free
Documentation License" in the Muse manual.  All Emacs Lisp code
contained in the manual may be used, distributed, and modified without
restriction.

The PNG images included in the examples/ directory may be used,
distributed, and modified without restriction.

The contents of the contrib/blosxom and contrib/pyblosxom directories
are available under the terms of other licenses.  Please consult the
header of each file for the appropriate license text.


NEWS
Emacs Muse NEWS --- History of user-visible changes    -*- outline -*-

* Changes in Muse 3.12

** New interactive function `muse-update-values': Call this after
changing muse-project-alist, in order to update various autogenerated
values.

** Tag attributes may now span multiple lines.

** Don't keep track of undo data when publishing.
This should yield a speed-up.

** Build system

*** Add support for the DESTDIR variable, which allows the destination
directory to be easily set.

** Muse Mode highlighting (lisp/muse-colors.el)

*** Display message when muse-colors-toggle-inline-images is called.

** ConTeXt publishing (lisp/muse-context.el)

*** Fix publishing bug in Windows.

** HTML publishing (lisp/muse-html.el)

*** Fix bug where nested class tags were not working.

*** Fix error that occurred when a class tag had no name element.
In this case, we do not publish the tag at all.

** LaTeX publishing (lisp/muse-latex.el)

*** Escape specials in image filenames properly
Previously, it was not possible to escape "/", "#", or "|" in image
filenames.  In order to make this work, be certain that you have the
following LaTeX code in your header.

\def\museincludegraphics{%
  \begingroup
  \catcode`\|=0
  \catcode`\\=12
  \catcode`\#=12
  \includegraphics[width=0.75\textwidth]
}

*** Escape specials in the title string in headers
In order to achieve this effect in custom headers, replace

  \title{<lisp>(muse-publishing-directive "title")</lisp>}

with

  \title{<lisp>(muse-publish-escape-specials-in-string
    (muse-publishing-directive "title") 'document)</lisp>}

in your headers.

*** Default to using UTF-8 rather than latin1 in headers.

*** Fix publishing bug in Windows.

*** Escape backslash as \textbackslash{} in monospace regions.

** Publish embedded LaTeX content to a PNG file (lisp/muse-latex2png.el)

*** Detect whether the tag ends at the end of a line.
If not, do not use "$$" to publish it.

** Muse Mode (lisp/muse-mode.el)

*** New option: `muse-insert-url-initial-input' specifies the initial text
to use when reading a URL.

*** Fix bug with browsing anchors that come after a link to them.

*** Fix bug involving filling and paragraphs next to headings.

*** Fill definition lists with two initial spaces in lines after the
definition list term.

** Project support (lisp/muse-project.el)

*** Fix issue that occurred when saving several files at once, where only
one of them would make it into the project file-alist.

*** If the PAGE argument to `muse-project-page-file' is nil, then return
the first directory of the project.

*** Fix bug where directory names without slashes in projects were not
being recognized.

** Publishing (lisp/muse-publish.el)

*** New tag <br> represents a line break when publishing.
This is supported in all publishing styles except for DocBook, which
has no notion of line breaks without stylesheet hacks.

*** Treat "---" as an mdash.
Before, it would publish as an mdash followed by a single dash.

*** Escape specials properly in footnotes in several styles.

*** Fix bug with publishing plain URLs.

*** Fix bad escaping of image links.

*** Save match adta in `muse-publish-classify-url'.

** Wiki-like behavior (lisp/muse-wiki.el)

*** Display <nop> tags correctly.

*** Publish <nop> tags correctly.

* Changes in Muse 3.11

** Update my example settings in examples/mwolson.

** Fix several bugs with setting muse-file-extension to something else.
If you have both muse-file-extension set to some string, and
muse-mode-auto-p set to non-nil, please set muse-mode-auto-p to nil
from now on.  This was a workaround that some people used to deal with
a bug that has now been fixed.

** Support for serving published Muse files with Blosxom.
See the end of the Blosxom Requirements section of the Muse manual for
details.

** The metadate plugin for PyBlosxom is now included with Muse
in contrib/pyblosxom/metadate.py.

** Compatibility fixes for Emacs21 and XEmacs

*** Use copy-tree instead of copy-alist.

*** Correctly require derived.el.

*** Deal with the lack of a `delete-and-extract-region' function in XEmacs
by making `muse-delete-and-extract-region'.

*** Fix XEmacs and Texinfo publishing bug.

*** Deal with lack of autoloads for the `man' function in some Emacs variants.

*** Fix XEmacs compilation error in muse-import-xml.el.

** Muse Mode highlighting (lisp/muse-colors.el)

*** Fix display bug where emphasis becomes unhighlighted when moving

muse-html.el
;;; muse-html.el --- Publish to HTML and XHTML.

;; Copyright (C) 2004, 2005  Free Software Foundation, Inc.

;; This file is not part of GNU Emacs.

;; This is free software; you can redistribute it and/or modify it under
;; the terms of the GNU General Public License as published by the Free
;; Software Foundation; either version 2, or (at your option) any later
;; version.
;;
;; This is distributed in the hope that it will be useful, but WITHOUT
;; ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
;; FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
;; for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;;; Contributors:

;; Zhiqiang Ye (yezq AT mail DOT cbi DOT pku DOT edu DOT cn) suggested
;; appending an 'encoding="..."' fragment to the first line of the
;; sample publishing header so that when editing the resulting XHTML
;; file, Emacs would use the proper encoding.

;;; Code:

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
;; Muse HTML Publishing
;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(require 'muse-publish)
(require 'muse-regexps)

(defgroup muse-html nil
  "Options controlling the behavior of Muse HTML publishing.
See `muse-html' for more information."
  :group 'muse-publish)

(defcustom muse-html-extension ".html"
  "Default file extension for publishing HTML files."
  :type 'string
  :group 'muse-html)

(defcustom muse-html-style-sheet
  "<style type=\"text/css\">
body {
  background: white; color: black;
  margin-left: 3%; margin-right: 7%;
}

p { margin-top: 1% }
p.verse { margin-left: 3% }

.example { margin-left: 3% }

h2 {
  margin-top: 25px;
  margin-bottom: 0px;
}
h3 { margin-bottom: 0px; }
    </style>"
  "Store your stylesheet definitions here.
This is used in `muse-html-header'.
You can put raw CSS in here or a <link> tag to an external stylesheet.
This text may contain <lisp> markup tags.

An example of using <link> is as follows.

<link rel=\"stylesheet\" type=\"text/css\" charset=\"utf-8\" media=\"all\" href=\"/default.css\">

If you are using XHTML, make sure to close the tag properly, as
shown in the following example.

<link rel=\"stylesheet\" type=\"text/css\" charset=\"utf-8\" media=\"all\" href=\"/default.css\" />"
  :type 'string
  :group 'muse-html)

(defcustom muse-html-header
  "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.0//EN\">
<html>
  <head>
    <title><lisp>
  (concat (muse-publishing-directive \"title\")
          (let ((author (muse-publishing-directive \"author\")))
            (if (not (string= author (user-full-name)))
                (concat \" (by \" author \")\"))))</lisp></title>
    <meta name=\"generator\" content=\"muse.el\">
    <meta http-equiv=\"<lisp>muse-html-meta-http-equiv</lisp>\"
          content=\"<lisp>muse-html-meta-content-type</lisp>\">
    <lisp>
      (let ((maintainer (muse-style-element :maintainer)))
        (when maintainer
          (concat \"<link rev=\\\"made\\\" href=\\\"\" maintainer \"\\\">\")))
    </lisp>
    <lisp>muse-html-style-sheet</lisp>
  </head>
  <body>
    <h1><lisp>
  (concat (muse-publishing-directive \"title\")
          (let ((author (muse-publishing-directive \"author\")))
            (if (not (string= author (user-full-name)))
                (concat \" (by \" author \")\"))))</lisp></h1>
    <!-- Page published by Emacs Muse begins here -->\n"
  "Header used for publishing HTML files."
  :type '(choice string file)
  :group 'muse-html)

(defcustom muse-html-footer "
<!-- Page published by Emacs Muse ends here -->
  </body>
</html>\n"
  "Footer used for publishing HTML files."
  :type '(choice string file)
  :group 'muse-html)

(defcustom muse-xhtml-header
  "<?xml version=\"1.0\" encoding=\"<lisp>
  (muse-html-encoding)</lisp>\"?>
<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Strict//EN\"
    \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd\">
<html xmlns=\"http://www.w3.org/1999/xhtml\">
  <head>
    <title><lisp>
  (concat (muse-publishing-directive \"title\")
          (let ((author (muse-publishing-directive \"author\")))
            (if (not (string= author (user-full-name)))
                (concat \" (by \" author \")\"))))</lisp></title>
    <meta name=\"generator\" content=\"muse.el\" />
    <meta http-equiv=\"<lisp>muse-html-meta-http-equiv</lisp>\"
          content=\"<lisp>muse-html-meta-content-type</lisp>\" />
    <lisp>
      (let ((maintainer (muse-style-element :maintainer)))
        (when maintainer
          (concat \"<link rev=\\\"made\\\" href=\\\"\" maintainer \"\\\" />\")))
    </lisp>
    <lisp>muse-html-style-sheet</lisp>
  </head>
  <body>
    <h1><lisp>
  (concat (muse-publishing-directive \"title\")
          (let ((author (muse-publishing-directive \"author\")))
            (if (not (string= author (user-full-name)))
                (concat \" (by \" author \")\"))))</lisp></h1>
    <!-- Page published by Emacs Muse begins here -->\n"
  "Header used for publishing XHTML files."
  :type '(choice string file)
  :group 'muse-html)

(defcustom muse-xhtml-footer "
<!-- Page published by Emacs Muse ends here -->
  </body>
</html>\n"
  "Footer used for publishing XHTML files."
  :type '(choice string file)
  :group 'muse-html)

(defcustom muse-html-anchor-on-word nil
  "When true, anchors surround the closest word. This allows you
to select them in a browser (i.e. for pasting), but has the
side-effect of marking up headers in multiple colors if your
header style is different from your link style."
  :type 'boolean
  :group 'muse-html)

(defcustom muse-html-table-attributes
  "class=\"muse-table\" border=\"2\" cellpadding=\"5\""
  "The attribute to be used with HTML <table> tags.
Note that since Muse supports direct insertion of HTML tags, you
can easily create any kind of table you want, as long as each
line begins at column 0 (to prevent it from being blockquoted).
To make such a table, use this idiom:

  <verbatim>
  <table>
    [... contents of my table, in raw HTML ...]
  </verbatim></table>

It may look strange to have the tags out of sequence, but this is
because the Muse verbatim tag is handled during a different pass
than the HTML table tag."
  :type 'string
  :group 'muse-html)

(defcustom muse-html-markup-regexps
  `(;; join together the parts of a list or table
    (10000 "</\\([oud]l\\)>\\s-*<\\1>\\s-*" 0 "")
    (10100 ,(concat "  </t\\(body\\|head\\|foot\\)>\\s-*</table>\\s-*"
                    "<table[^>]*>\\s-*<t\\1>\n") 0 "")
    (10200 "</table>\\s-*<table[^>]*>\n" 0 "")

    ;; beginning of doc, end of doc, or plain paragraph separator
    (10300 ,(concat "\\(\n</\\(blockquote\\|center\\)>\\)?"
                    "\\(?:\n\\(["
                    muse-regexp-blank
                    "]*\n\\)+\\|\\`\\s-*\\|\\s-*\\'\\)"
                    "\\(<\\(blockquote\\|center\\)>\n\\)?")
           0 muse-html-markup-paragraph))
  "List of markup rules for publishing a Muse page to HTML.
For more on the structure of this list, see `muse-publish-markup-regexps'."
  :type '(repeat (choice
                  (list :tag "Markup rule"
                        integer
                        (choice regexp symbol)
                        integer
                        (choice string function symbol))
                  function))
  :group 'muse-html)

(defcustom muse-html-markup-functions
  '((anchor   . muse-html-markup-anchor)
    (table    . muse-html-markup-table)
    (footnote . muse-html-markup-footnote))
  "An alist of style types to custom functions for that kind of text.
For more on the structure of this list, see
`muse-publish-markup-functions'."
  :type '(alist :key-type symbol :value-type function)
  :group 'muse-html)

(defcustom muse-html-markup-strings
  '((image-with-desc . "<img src=\"%s\" alt=\"%s\">")
    (image-link      . "<img src=\"%s\">")
    (url-with-image  . "<a class=\"image-link\" href=\"%s\"><img src=\"%s\"></a>")
    (url-link        . "<a href=\"%s\">%s</a>")
    (email-addr      . "<a href=\"mailto:%s\">%s</a>")
    (emdash          . " &mdash; ")
    (rule            . "<hr>")
    (fn-sep          . "<hr>\n")
    (enddots         . "....")
    (dots            . "...")
    (section         . "<h2>")
    (section-end     . "</h2>")
    (subsection      . "<h3>")
    (subsection-end  . "</h3>")
    (subsubsection   . "<h4>")
    (subsubsection-end . "</h4>")
    (begin-underline . "<u>")
    (end-underline   . "</u>")
    (begin-literal   . "<code>")
    (end-literal     . "</code>")
    (begin-emph      . "<em>")
    (end-emph        . "</em>")
    (begin-more-emph . "<strong>")
    (end-more-emph   . "</strong>")
    (begin-most-emph . "<strong><em>")
    (end-most-emph   . "</em></strong>")
    (begin-verse     . "<p class=\"verse\">\n")
    (verse-space     . "&nbsp;&nbsp;")
    (end-verse-line  . "<br>")
    (last-stanza-end . "<br>")
    (empty-verse-line . "<br>")
    (end-verse       . "</p>")
    (begin-example   . "<pre class=\"example\">")
    (end-example     . "</pre>")
    (begin-center    . "<center>\n")
    (end-center      . "\n</center>")
    (begin-quote     . "<blockquote>\n")
    (end-quote       . "\n</blockquote>")
    (begin-uli       . "<ul>\n<li>")
    (end-uli         . "</li>\n</ul>")
    (begin-oli       . "<ol>\n<li>")
    (end-oli         . "</li>\n</ol>")
    (begin-ddt       . "<dl>\n<dt><strong>")
    (start-dde       . "</strong></dt>\n<dd>")
    (end-ddt         . "</dd>\n</dl>"))
  "Strings used for marking up text as HTML.
These cover the most basic kinds of markup, the handling of which
differs little between the various styles."
  :type '(alist :key-type symbol :value-type string)
  :group 'muse-html)

(defcustom muse-xhtml-markup-strings
  '((image-with-desc . "<img src=\"%s\" alt=\"%s\" />")
    (image-link      . "<img src=\"%s\" alt=\"\" />")
    (url-with-image  . "<a class=\"image-link\" href=\"%s\"><img src=\"%s\" alt=\"\" /></a>")
    (rule            . "<hr />")
    (fn-sep          . "<hr />\n")
    (begin-underline . "<span style=\"text-decoration: underline;\">\n")
    (end-underline   . "</span>")
    (begin-center    . "<span style=\"text-align: center;\">\n")
    (end-verse-line  . "<br />")
    (last-stanza-end . "<br />")
    (empty-verse-line . "<br />")
    (end-center      . "\n</span>"))
  "Strings used for marking up text as XHTML.
These cover the most basic kinds of markup, the handling of which
differs little between the various styles.

If a markup rule is not found here, `muse-html-markup-strings' is
searched."
  :type '(alist :key-type symbol :value-type string)
  :group 'muse-html)

(defcustom muse-html-markup-tags
  '(("class" t t muse-html-class-tag))
  "A list of tag specifications, for specially marking up HTML."
  :type '(repeat (list (string :tag "Markup tag")
                       (boolean :tag "Expect closing tag" :value t)
                       (boolean :tag "Parse attributes" :value nil)
                       function))
  :group 'muse-html)

(defcustom muse-html-markup-specials
  '((?\" . "&quot;")
    (?\< . "&lt;")
    (?\> . "&gt;")
    (?\& . "&amp;"))
  "A table of characters which must be represented specially."
  :type '(alist :key-type character :value-type string)
  :group 'muse-html)

(defcustom muse-html-meta-http-equiv "Content-Type"
  "The http-equiv attribute used for the HTML <meta> tag."
  :type 'string
  :group 'muse-html)

(defcustom muse-html-meta-content-type "text/html"
  "The content type used for the HTML <meta> tag.
If you are striving for XHTML 1.1 compliance, you may want to
change this to \"application/xhtml+xml\"."
  :type 'string
  :group 'muse-html)

(defcustom muse-html-meta-content-encoding (if (featurep 'mule)
                                              'detect
                                            "iso-8859-1")
  "The charset to append to the HTML <meta> tag.
If set to the symbol 'detect, use `muse-html-encoding-map' to try
and determine the HTML charset from emacs's coding.  If set to a
string, this string will be used to force a particular charset"
  :type '(choice string symbol)
  :group 'muse-html)

(defcustom muse-html-charset-default "iso-8859-1"
  "The default HTML meta charset to use if no translation is found in
`muse-html-encoding-map'."
  :type 'string
  :group 'muse-html)

(defcustom muse-html-encoding-default 'iso-8859-1
  "The default Emacs buffer encoding to use in published files.
This will be used if no special characters are found."
  :type 'symbol
  :group 'muse-html)

(defcustom muse-html-encoding-map
  '((iso-2022-jp        . "iso-2022-jp")
    (utf-8              . "utf-8")
    (japanese-iso-8bit  . "euc-jp")
    (chinese-big5       . "big5")
    (mule-utf-8         . "utf-8")
    (chinese-iso-8bit   . "gb2312")
    (chinese-gbk        . "gbk"))
  "An alist mapping emacs coding systems to appropriate HTML charsets.
Use the base name of the coding system (i.e. without the -unix)."
  :type '(alist :key-type coding-system :value-type string)
  :group 'muse-html)

(defun muse-html-transform-content-type (content-type)
  "Using `muse-html-encoding-map', try and resolve an emacs coding
system to an associated HTML coding system. If no match is found,
`muse-html-charset-default' is used instead."
  (let ((match (assoc (coding-system-base content-type)
                      muse-html-encoding-map)))
    (if match
        (cdr match)
      muse-html-charset-default)))

(defun muse-html-insert-anchor (anchor)
  "Insert an anchor, either around the word at point, or within a tag."
  (skip-chars-forward muse-regexp-space)
  (if (looking-at "<\\([^ />]+\\)>")
      (let ((tag (match-string 1)))
        (goto-char (match-end 0))
        (insert "<a name=\"" anchor "\" id=\"" anchor "\">")
        (when muse-html-anchor-on-word
          (or (and (search-forward (format "</%s>" tag)
                                   (line-end-position) t)
                   (goto-char (match-beginning 0)))
              (forward-word 1)))
        (insert "</a>"))
    (insert "<a name=\"" anchor "\" id=\"" anchor "\">")
    (when muse-html-anchor-on-word
      (forward-word 1))
    (insert "</a>")))

(defun muse-html-markup-paragraph ()
  (let ((end (copy-marker (match-end 0) t)))
    (goto-char (match-beginning 0))
    (insert
     (or (save-excursion
           (save-match-data
             (and (re-search-backward "<\\(/?\\)\\(p\\|div\\)[ >]"
                                      nil t)
                  (not (string-equal (match-string 1) "/"))
                  (if (string-equal (match-string 2) "p")
                      "</p>"
                    "</div>"))))
         ""))
    (goto-char end))
  (cond
   ((eobp)
    (unless (bolp)
      (insert "\n")))
   ((eq (char-after) ?\<)
    (cond
     ((looking-at "<\\(em\\|strong\\|code\\|span\\)[ >]")
      (insert "<p>"))
     ((looking-at "<a ")
      (if (looking-at "<a[^>]+><img")
          (insert "<div class=\"image-link\">")
        (insert "<p>")))
     ((looking-at "<img[ >]")
      (insert "<div class=\"image-link\">"))))
   ((muse-looking-back "\\(</h[1-4]>\\|<hr>\\)\n\n")
    (insert "<p class=\"first\">"))
   ((muse-looking-back "<\\(blockquote\\|center\\)>\n")
    (insert "<p class=\"quoted\">"))
   (t
    (insert "<p>"))))

(defun muse-html-markup-anchor ()
  (save-match-data
    (muse-html-insert-anchor (match-string 1))) "")

(defun muse-html-escape-string (str)
  "Convert to character entities any non-alphanumeric characters
outside a few punctuation symbols, that risk being misinterpreted
if not escaped."
  (when str
    (let (pos code len)
      (save-match-data
        (while (setq pos (string-match (concat "[^-"
                                               muse-regexp-alnum
                                               "/:._=@\\?~#]")
                                       str pos))
          (setq code (int-to-string
                      (cond ((fboundp 'char-to-ucs)
                             (char-to-ucs (aref str pos)))
                            ((fboundp 'char-to-int)
                             (char-to-int (aref str pos)))
                            (t (aref str pos))))
                len (length code)
                         str (concat (substring str 0 pos)
                                     "&#" code ";"
                                     (when (< pos (length str))
                                       (substring str (1+ pos) nil)))
                pos (+ 3 len pos)))
        str))))

(defun muse-html-markup-footnote ()
  (if (/= (line-beginning-position) (match-beginning 0))
      "<sup><a name=\"fnr.\\1\" href=\"#fn.\\1\">\\1</a></sup>"
    (prog1
        "<p class=\"footnote\"><a name=\"fn.\\1\" href=\"#fnr.\\1\">\\1.</a>"
      (save-excursion
        (save-match-data
          (let* ((beg (goto-char (match-end 0)))
                 (end (and (search-forward "\n\n" nil t)
                           (prog1
                               (copy-marker (match-beginning 0))
                             (goto-char beg)))))
            (while (re-search-forward (concat "^["
                                              muse-regexp-blank
                                              "]+\\([^\n]\\)")
                                      end t)
              (replace-match "\\1" t))))))))

(defun muse-html-markup-table ()
  (let* ((str (prog1
                  (match-string 1)
                (delete-region (match-beginning 0) (match-end 0))))
         (fields (split-string str "\\s-*|+\\s-*"))
         (type (and (string-match "\\s-*\\(|+\\)\\s-*" str)
                    (length (match-string 1 str))))
         (part (cond ((= type 1) "tbody")
                     ((= type 2) "thead")
                     ((= type 3) "tfoot")))
         (col (cond ((= type 1) "td")
                    ((= type 2) "th")
                    ((= type 3) "td")))
         field)
    (insert "<table " muse-html-table-attributes ">\n"
            "  <" part ">\n"
            "    <tr>\n")
    (dolist (field fields)
      (insert "      <" col ">" field "</" col ">\n"))
    (insert "    </tr>\n"
            "  </" part ">\n"
            "</table>\n")))

;; Handling of tags for HTML

(defun muse-html-insert-contents (depth)
  (let ((max-depth (or depth 2))
        (index 1)
        base contents l)
    (save-excursion
      (goto-char (point-min))
      (search-forward "Page published by Emacs Muse begins here" nil t)
      (catch 'done
        (while (re-search-forward "^<h\\([0-9]+\\)>\\(.+?\\)</h\\1>" nil t)
          (unless (get-text-property (point) 'read-only)
            (setq l (1- (string-to-number (match-string 1))))
            (if (null base)
                (setq base l)
              (if (< l base)
                  (throw 'done t)))
            (when (<= l max-depth)
              (setq contents (cons (cons l (match-string-no-properties 2))
                                   contents))
              (goto-char (match-beginning 2))
              (muse-html-insert-anchor (concat "sec" (int-to-string index)))
              (setq index (1+ index)))))))
    (setq index 1 contents (reverse contents))
    (let ((depth 1) (sub-open 0) (p (point)))
      (insert "<dl class=\"contents\">\n")
      (while contents
        (insert "<dt class=\"contents\">\n")
        (insert "<a href=\"#sec" (int-to-string index) "\">"
                (muse-publish-strip-tags (cdar contents))
                "</a>\n")
        (setq index (1+ index))
        (insert "</dt>\n")
        (setq depth (caar contents)
              contents (cdr contents))
        (if contents
            (cond
             ((< (caar contents) depth)
              (let ((idx (caar contents)))
                (while (< idx depth)
                  (insert "</dl>\n</dd>\n")
                  (setq sub-open (1- sub-open)
                        idx (1+ idx)))))
             ((> (caar contents) depth) ; can't jump more than one ahead
              (insert "<dd>\n<dl class=\"contents\">\n")
              (setq sub-open (1+ sub-open))))))
      (while (> sub-open 0)
        (insert "</dl>\n</dd>\n")
        (setq sub-open (1- sub-open)))
      (insert "</dl>\n")
      (muse-publish-mark-read-only p (point)))))

(defun muse-html-class-tag (beg end attrs)
  (goto-char beg)
  (insert "<span class=\"" (cdr (assoc "name" attrs)) "\">")
  (goto-char end)
  (insert "</span>"))

;; Register the Muse HTML Publisher

(defun muse-html-browse-file (file)
  (browse-url (concat "file:" file)))

(defun muse-html-encoding ()
  (if (stringp muse-html-meta-content-encoding)
      muse-html-meta-content-encoding
    (muse-html-transform-content-type
     (or buffer-file-coding-system
         muse-html-encoding-default))))

(defun muse-html-prepare-buffer ()
  (set (make-local-variable 'muse-publish-url-transforms)
       (cons 'muse-html-escape-string muse-publish-url-transforms))
  (make-local-variable 'muse-html-meta-http-equiv)
  (set (make-local-variable 'muse-html-meta-content-type)
       (concat muse-html-meta-content-type "; charset="
               (muse-html-encoding))))

(defun muse-html-finalize-buffer ()
  (when muse-publish-generate-contents
    (goto-char (car muse-publish-generate-contents))
    (muse-html-insert-contents (cdr muse-publish-generate-contents)))
  (when (memq buffer-file-coding-system '(no-conversion undecided-unix))
    ;; make it agree with the default charset
    (setq buffer-file-coding-system muse-html-encoding-default)))

(unless (assoc "html" muse-publishing-styles)
  (muse-define-style "html"
                     :suffix    'muse-html-extension
                     :regexps   'muse-html-markup-regexps
                     :functions 'muse-html-markup-functions
                     :strings   'muse-html-markup-strings
                     :tags      'muse-html-markup-tags
                     :specials  'muse-html-markup-specials
                     :before    'muse-html-prepare-buffer
                     :after     'muse-html-finalize-buffer
                     :header    'muse-html-header
                     :footer    'muse-html-footer
                     :browser   'muse-html-browse-file)

  (muse-derive-style "xhtml" "html"
                     :strings   'muse-xhtml-markup-strings
                     :header    'muse-xhtml-header
                     :footer    'muse-xhtml-footer))

(provide 'muse-html)

;;; muse-html.el ends here
